### ISSUE TITLE
ClusteredEventBus.removeRegistration does not call completionHandler on failure

### ISSUE THREAD

**Issue by rvansa** [eclipse-vertx/vert.x#3638]
### Version

3.9.3, present in last 3.9, it seems to be fixed in master

### Context

Closing clustered Vert.x with **custom-provided Infinispan cluster manager**. The CM is closed in `VertxInternal.addCloseHook(...)` handler.

In method `ClusteredEventBus.removeSub` the `completionHandler` is invoked if and only if the `subs.remove(...)` succeeds; if that fails only an error is logged but the handler is not invoked. Consequently the handler for `vertx.close(...)` is never called.

In my case the call fails because the cluster manager is already closed when the eventbus is being unregistered.

### Steps to reproduce

```java
DefaultCacheManager cacheManager = ...
options.setClusterManager(new InfinispanClusterManager(cacheManager));
Vertx.clusteredVertx(options, result -> {
   if (result.failed()) {
      /* ... */
   }
   Vertx vertx = result.result();
   /* ... */
   ((VertxInternal) vertx).addCloseHook(handler -> {
      cacheManager.stop();
      handler.handle(Future.succeededFuture());
   });
   /* ... */
});
```

### Extra

If there is a way to properly close the CM upon Vert.x close please let me know.

**Comment by tsegismont** [comment]
Hi @rvansa . The `close` handler should be invoked even in case of failure. I'll look into this and fix any issue I may find.

As for closing the `cacheManager` and `vertx` properly, how about removing your close hook and close the `cacheManager` in the callback?

```java
DefaultCacheManager cacheManager = ...
options.setClusterManager(new InfinispanClusterManager(cacheManager));
Vertx.clusteredVertx(options, result -> {
   if (result.failed()) {
      /* ... */
   }
   Vertx vertx = result.result();
   /* ... */
});

vertx.close(v -> {
  cacheManager.stop();
});
```

**Comment by tsegismont** [comment]
@rvansa this is the part of the code you were talking about, correct?

https://github.com/eclipse-vertx/vert.x/blob/baf77ec2e1931965758a0b53b6da6b3473115b56/src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java#L412-L414

I reckon there is a problem here in 3.9 But it shouldn't impact shutdown because `EventBus` unregisters all handlers without taking the result into account.

Can you show me a complete stacktrace?

**Comment by rvansa** [comment]
@tsegismont Yes, I've ended up closing the CM in the `vertx.close(...)` handler - that works. However it means that I need to centralize closing Vert.x in my application and do a couple of casts to get the cache manager instance, instead of just tying CM lifecycle to Vert.x when I create it:
```java
ClusterManager clusterManager = ((VertxInternal) vertx).getClusterManager();
DefaultCacheManager cacheManager = (DefaultCacheManager) ((InfinispanClusterManager) clusterManager).getCacheContainer();
vertx.close(result -> {
   try {
      cacheManager.close();
   } catch (IOException e) {
      log.error("Failed to close Infinispan cache manager", e);
   } finally {
      if (handler != null) {
          handler.handle(result);
      }
   }
});
```

Below is the stacktrace when the `removeSub` is called. 
```
removeSub:409, ClusteredEventBus (io.vertx.core.eventbus.impl.clustered)
removeRegistration:203, ClusteredEventBus (io.vertx.core.eventbus.impl.clustered)
removeRegistration:289, EventBusImpl (io.vertx.core.eventbus.impl)
doUnregister:185, HandlerRegistration (io.vertx.core.eventbus.impl)
unregister:137, HandlerRegistration (io.vertx.core.eventbus.impl)
close:617, EventBusImpl$HandlerEntry (io.vertx.core.eventbus.impl)
run:105, CloseHooks (io.vertx.core.impl)
runCloseHooks:171, ContextImpl (io.vertx.core.impl)
runCloseHooks:23, EventLoopContext (io.vertx.core.impl)
lambda$null$5:703, DeploymentManager$DeploymentImpl (io.vertx.core.impl)
handle:-1, 729548303 (io.vertx.core.impl.DeploymentManager$DeploymentImpl$$Lambda$980)
dispatch:105, FutureImpl (io.vertx.core.impl)
tryComplete:150, FutureImpl (io.vertx.core.impl)
tryComplete:157, FutureImpl (io.vertx.core.impl)
complete:118, FutureImpl (io.vertx.core.impl)
lambda$stop$3:111, ControllerServer (io.hyperfoil.clustering)
handle:-1, 1521968818 (io.hyperfoil.clustering.ControllerServer$$Lambda$985)
lambda$executeCloseDone$11:507, HttpServerImpl (io.vertx.core.http.impl)
handle:-1, 347811196 (io.vertx.core.http.impl.HttpServerImpl$$Lambda$987)
executeTask:366, ContextImpl (io.vertx.core.impl)
lambda$executeAsync$0:38, EventLoopContext (io.vertx.core.impl)
run:-1, 955059858 (io.vertx.core.impl.EventLoopContext$$Lambda$135)
safeExecute:164, AbstractEventExecutor (io.netty.util.concurrent)
runAllTasks:472, SingleThreadEventExecutor (io.netty.util.concurrent)
run:500, NioEventLoop (io.netty.channel.nio)
run:989, SingleThreadEventExecutor$4 (io.netty.util.concurrent)
run:74, ThreadExecutorMap$2 (io.netty.util.internal)
run:30, FastThreadLocalRunnable (io.netty.util.concurrent)
run:834, Thread (java.lang)
```

**Comment by tsegismont** [comment]
@rvansa thanks for the stacktrace, it helped. So this happens when a verticle is undeployed. I'll fix this case. I checked that indeed it is fixed in `master`.

**Comment by tsegismont** [comment]
Closed in #3659
